xblock_skytap/public/js/src/skytap.js

```diff
@@ -34,13 +34,32 @@ function SkytapXBlock(runtime, element) {
         launchXHR = $.post(handlerUrl, JSON.stringify(keyboardLayout))
             .success(function(response) {
                 var url = response.sharing_portal_url;
-                var sharingPortal = window.open(url, '_blank');
-                if (sharingPortal === null) {
-                    alert(
-                        gettext("The browser's popup blocker prevented the exercise environment from being launched.")
-                    );
+
+                /*
+                 The standard behaviour is to open the exercise environment in a new tab using a popup,
+                 to allow the user to keep the course tab open too. However on iOS devices for example,
+                 popups are blocked by default and the user is not informed that the popup failed to open.
+                 Therefore for iOS devices a redirect is used instead. For consistency this is done for all
+                 mobile devices.
+                 */
+                var isiOS = navigator.userAgent.match(/(iPod|iPhone|iPad)/i);
+                var isAndroid = navigator.userAgent.match(/(android)/i);
+                var isWindows = navigator.userAgent.match(/(Windows Phone|iemobile)/i);
+                if (isiOS || isAndroid || isWindows) {
+                    // simply redirect for mobile devices
+                    window.location = url;
                 } else {
-                    sharingPortal.focus();
+                    // desktop browsers offer an easy way to allow the popup so being blocked is ok
+                    var sharingPortal = window.open(url, '_blank');
+                    if (sharingPortal == null) {
+                        // use '==' instead of '===' to test for both null (from desktop browsers blocking the popup)
+                        // and also undefined (from mobile Safari refusing to show the popup)
```

Comment:
@carlio Thanks for adding this comment to explain what's going on! :) Can we change the code to test for `null` and `undefined` *explicitly* (I'd still like to keep this comment, though)?